src/dispatch.rs

```diff
@@ -83,7 +84,7 @@ struct PutRequest {
 }
 
 struct Dispatcher {
-    connections: HashMap<String, IdbStore>,
+    connections: HashMap<String, Box<dyn Store>>,
```

Comment:
It seems like dynamic dispatch is vaguely discouraged in Rust, so I was trying in that thing I sent in Slack to find a way to not use it*.

I do not think that "idiomatic Rust" should be a goal for this project right now, but at the same time it's interesting to explore tradeoffs to learn the language.

Anyway, two questions:

1. Does async-trait end up compiling a `dyn Trait` result?
2. In TypeScript, `async foo() -> Bar` is another way to say `Foo() -> Promise<Bar>`. The functions both end up with the same type (the second one). The only difference is that inside the first you can use `await`. I wonder if it's the same in Rust. Can we avoid async-trait (and thereby have option to use static dispatch) by phrasing the trait with return type Future, but still *implement* the trait using async/await.

* OTOH, static dispatch is associated with increased code size which is the opposite of what we want **.

** OTOOH, in our case we'll only have one implementation compiled into the binary at a time so maybe we wouldn't see a size increase.

Reply:
> It seems like dynamic dispatch is vaguely discouraged in Rust, so I was trying in that thing I sent in Slack to find a way to not use it*.

In a quick reading, I do see folks talking about not using it, but just to avoid the small overhead that calling through a C++ vtable introduces. That is, I don't think that's it's discouraged b/c of Rust, but just because you *can* use generics to squeak out that extra .00001% of performance if you wish. Is there a further reason I missed?

> I do not think that "idiomatic Rust" should be a goal for this project right now, but at the same time it's interesting to explore tradeoffs to learn the language.
> 
> Anyway, two questions:
> 
> 1. Does async-trait end up compiling a `dyn Trait` result?

https://docs.rs/async-trait/0.1.36/async_trait/#explanation suggests that methods in async traits return a `Pin<Box<dyn Future + Send + 'async>>` (I opted out of the `Send` part via `#[async_trait(?Send)]` since IdbDatabase doens't support that).

I don't know if that answers your question though, does it? I'm cargo culting far more than I'm comfortable with because I'm not grokking much of what's going on yet. :/

> 2. In TypeScript, `async foo() -> Bar` is another way to say `Foo() -> Promise<Bar>`. The functions both end up with the same type (the second one). The only difference is that inside the first you can use `await`. I wonder if it's the same in Rust. Can we avoid async-trait (and thereby have option to use static dispatch) by phrasing the trait with return type Future, but still _implement_ the trait using async/await.

I don't understand yet how using async-trait prevents us from doing static dispatch here. I think that we *could* define `dispatch_loop` and `Dispatcher` as generics on a type implementing a trait, and then instantiate that once in each of our build products based on the Store implementation we want. Does the `dyn Future` in async-trait preclude that? In either case, especially at a level so close to mapping to I/O, I don't think a bit of indirection overhead will ever show up in a profile unlike how it might in some big in-memory graph algorithm.

> * OTOH, static dispatch is associated with increased code size which is the opposite of what we want **.
> 
> ** OTOOH, in our case we'll only have one implementation compiled into the binary at a time so maybe we wouldn't see a size increase.